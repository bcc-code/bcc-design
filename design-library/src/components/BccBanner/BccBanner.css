@layer components {
    .bcc-banner {
        background: var(--bcc-banner-bg, var(--brand-600));
        @apply p-4
    }

    .bcc-banner-header {
        @apply flex items-center justify-between text-white mb-4;
    }

    .bcc-banner-icon {
        @apply h-6 w-6 mr-3;
    }

    .bcc-banner-title {
        @apply text-heading-base flex-1;
    }

    .bcc-banner-close-button {
        @apply p-2 -mt-1 -mb-1 -mr-2;
    }
    .bcc-banner-close-button:active {
        @apply scale-90;
    }

    .bcc-banner-close-icon {
        @apply h-5 w-5;
    }

    .bcc-banner-default {
        @apply text-white;
    }

    .bcc-banner-default .bcc-banner-actions {
        @apply mt-3 flex flex-wrap sm:flex-nowrap gap-3;
    }

    .bcc-banner-default .bcc-banner-actions .bcc-button {
        @apply w-full;
    }

    .bcc-banner-default .bcc-button:not(:first-child) {
        --text-button-secondary: white;
        --border-button-secondary: var(--brand-500);
        --bg-button-secondary: var(--brand-700);
        --bg-button-secondary-hover: var(--brand-700);
    }

    .bcc-banner-notification {
        @apply bg-white p-3 rounded-xl;
    }

    .bcc-banner-notification .bcc-banner-actions {
        @apply flex items-center justify-end gap-3;
    }

    .bcc-banner-notification .bcc-banner-actions .bcc-button {
        @apply  min-w-min w-24;
    }
}