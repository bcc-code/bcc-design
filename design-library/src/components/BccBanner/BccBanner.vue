<script setup lang="ts">
import { CloseIcon } from "@bcc-code/icons-vue";
import type { Component } from "vue";
import BccButton from "../BccButton/BccButton.vue";

type Action = {
  label: string;
  icon?: string | Component | Function;
  onClick?: () => void;
};

type Props = {
  title: string;
  variant?: keyof typeof variants;
  icon?: string | Component | Function;
  closeButton?: boolean;
  closeOnAction?: boolean;
  actions?: Action[];
  open?: boolean;
};

const props = withDefaults(defineProps<Props>(), {
  variant: "default",
  closeButton: true,
  closeOnAction: true,
  open: true,
});

const emit = defineEmits<{
  close: [action?: Action];
}>();

const variants = {
  default: "bcc-banner-default",
  notification: "bcc-banner-notification",
};

function actionClicked(action: Action) {
  if (action.onClick) {
    action.onClick();
  }

  if (props.closeOnAction) {
    emit("close", action);
  }
}
</script>

<template>
  <Transition name="bcc-close">
    <div class="bcc-banner" v-if="open">
      <div class="bcc-banner-header">
        <component v-if="icon" :is="icon" class="bcc-banner-icon" />
        <h5 class="bcc-banner-title">{{ title }}</h5>
        <button
          @click.prevent="emit('close')"
          v-if="closeButton"
          aria-label="Close banner"
          class="bcc-banner-close-button"
        >
          <CloseIcon class="bcc-banner-close-icon" />
        </button>
      </div>
      <div :class="variants[variant]">
        <slot></slot>

        <div v-if="actions && actions.length" class="bcc-banner-actions">
          <template v-for="(action, index) in actions" :key="index">
            <BccButton
              :icon="action.icon"
              icon-right
              size="sm"
              :variant="variant === 'notification' ? 'tertiary' : 'secondary'"
              @click="actionClicked(action)"
            >
              {{ action.label }}
            </BccButton>
          </template>
        </div>
      </div>
    </div>
  </Transition>
</template>
