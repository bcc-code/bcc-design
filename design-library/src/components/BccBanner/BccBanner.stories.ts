import {
  BccFillIcon,
  CircleNotificationsFillIcon,
  NotificationsOffFillIcon
} from "@bcc-code/icons-vue";
import BccBanner from "./BccBanner.vue";

import type { Meta, StoryFn } from "@storybook/vue3";

export default {
  title: "Components/BccBanner",
  component: BccBanner,
  argTypes: {
    variant: {
      options: ["default", "notification"],
      control: { type: "radio" },
    },
    title: {
      description: "Title of the banner",
    },
    open: {
      description: "Show the banner",
      control: { type: "boolean" },
    },
    closeOnAction: {
      description: "Close the banner when an action is clicked",
      control: { type: "boolean" },
    },
    slotDefault: {
      name: "default slot",
      description: "The button text",
    },
  },
} as Meta<typeof BccBanner>;

const Template: StoryFn<typeof BccBanner> = (args) => ({
  components: { BccBanner },
  setup() {
    function close() {
      args.open = false;
      setTimeout(() => {
        args.open = true;
      }, 1000);
    }
    return { args, close };
  },
  template: `
    <BccBanner v-bind="args" @close="close">
      <small>SLOTTED CONTENT</small>
      <p>Welcome to the jungle. Hope you have fun!</p>
    </BccBanner>
  `,
});

export const Example = Template.bind({});
Example.args = {
  variant: "default",
  icon: BccFillIcon,
  title: "Banner title",
  open: true,
  closeOnAction: true,
  actions: [
    {
      label: "Yes",
      icon: CircleNotificationsFillIcon,
      onClick: () => {
        alert("Yes");
      },
    },
    {
      label: "No",
      icon: NotificationsOffFillIcon,
      onClick: () => {
        alert("No");
      },
    },
  ],
};
