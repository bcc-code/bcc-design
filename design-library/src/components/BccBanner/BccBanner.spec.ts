import { describe, expect, it, vi } from "vitest";

import { mount } from "@vue/test-utils";
import BccBanner from "./BccBanner.vue";

describe("BccBanner", () => {
  it("renders an banner", () => {
    const wrapper = mount(BccBanner, {
      props: {
        title: "Banner title",
        actions: [
          {
            label: "OK",
          },
        ],
      },
      slots: {
        default: "Banner content",
      },
    });

    expect(wrapper.text()).toContain("Banner title");
    expect(wrapper.text()).toContain("Banner content");
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("can close", async () => {
    const wrapper = mount(BccBanner, {
      props: { title: "Banner title", open: true, closeButton: true },
      slots: {
        default: "Banner content",
      },
    });

    await wrapper.find(".bcc-banner-close-button").trigger("click");

    expect(wrapper.emitted("close")?.length).toBe(1);
  });

  it("can close via action", async () => {
    const actionClick = vi.fn(() => 1);
    const wrapper = mount(BccBanner, {
      props: {
        title: "Banner title",
        open: true,
        closeButton: true,
        actions: [
          {
            label: "OK",
            onClick: actionClick,
          },
        ],
      },
      slots: {
        default: "Banner content",
      },
    });

    await wrapper.find(".bcc-banner-actions .bcc-button").trigger("click");

    expect(actionClick).toHaveBeenCalled();

    expect(wrapper.emitted("close")?.length).toBe(1);
  });
});
