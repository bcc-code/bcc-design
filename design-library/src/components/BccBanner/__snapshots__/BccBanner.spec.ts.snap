// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BccBanner > renders an banner 1`] = `
"<transition-stub name=\\"bcc-close\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
  <div class=\\"bcc-banner\\">
    <div class=\\"bcc-banner-header\\">
      <!--v-if-->
      <h5 class=\\"bcc-banner-title\\">Banner title</h5><button aria-label=\\"Close banner\\" class=\\"bcc-banner-close-button\\"><svg aria-hidden=\\"true\\" viewBox=\\"0 -960 960 960\\" fill=\\"currentColor\\" class=\\"bcc-banner-close-icon\\">
          <path d=\\"M480-416.348 287.826-224.174Q275.152-211.5 256-211.5t-31.826-12.674Q211.5-236.848 211.5-256t12.674-31.826L416.348-480 224.174-672.174Q211.5-684.848 211.5-704t12.674-31.826Q236.848-748.5 256-748.5t31.826 12.674L480-543.652l192.174-192.174Q684.848-748.5 704-748.5t31.826 12.674Q748.5-723.152 748.5-704t-12.674 31.826L543.652-480l192.174 192.174Q748.5-275.152 748.5-256t-12.674 31.826Q723.152-211.5 704-211.5t-31.826-12.674L480-416.348Z\\"></path>
        </svg></button>
    </div>
    <div class=\\"bcc-banner-default\\">Banner content<div class=\\"bcc-banner-actions\\"><button class=\\"bcc-button bcc-button-sm bcc-button-secondary bcc-button-center\\">
          <!--v-if--><span class=\\"order-1\\">OK</span>
        </button></div>
    </div>
  </div>
</transition-stub>"
`;
